Firestore/core/src/firebase/firestore/model/field_value.h

```diff
@@ -128,6 +133,32 @@ class FieldValue {
     return ObjectValue{object_value_};
   }
 
+  /**
+   * Returns a FieldValue with the field at the named path set to value.
+   *
+   * @param field_path The field path to set.
+   * @param value The value to set.
+   * @return A new FieldValue with the field set.
+   */
+  FieldValue Set(FieldPath field_path, FieldValue value) const;
+
+  /**
+   * Returns a FieldValue with the field path deleted. If there is no field at
+   * the specified path nothing is changed.
+   *
+   * @param field_path The field path to remove
+   * @return A new FieldValue with the field path removed.
+   */
+  FieldValue Delete(FieldPath field_path) const;
+
+  /**
+   * Returns the value at the given path or absl::nullopt.
+   *
+   * @param field_path the path to search
+   * @return The value at the path or absl::nullopt if it doesn't exist.
+   */
+  absl::optional<FieldValue> Get(FieldPath field_path) const;
```

Comment:
This returns a copy of the value. Is copying intentional?

Reply:
It has to be a copy since "optional < FieldValue >  -> FieldValue -> map of FieldValue" everything is stack-allocated and nowhere in the data structure supports shared objects. This is expected to change once the immutable container type is ported, where the identical nodes will be shared across immutable container objects.